migrator/cold_migrate.go

```diff
@@ -0,0 +1,405 @@
+	"fmt"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/pouchcontainer/d2p-migrator/ctrd"
+	"github.com/pouchcontainer/d2p-migrator/docker"
+	"github.com/pouchcontainer/d2p-migrator/pouch"
+	"github.com/pouchcontainer/d2p-migrator/pouch/convertor"
+	localtypes "github.com/pouchcontainer/d2p-migrator/pouch/types"
+	"github.com/pouchcontainer/d2p-migrator/utils"
+
+	pouchtypes "github.com/alibaba/pouch/apis/types"
+	"github.com/containerd/containerd/errdefs"
+	dockerclient "github.com/docker/engine-api/client"
+	"github.com/sirupsen/logrus"
+)
+
+func init() {
+	Register("cold-migrate", NewColdMigrator)
+}
+
+// Actions that coldMigrator migration does.
+// 0. Install containerd1.0.3
+// 1. Pull Images
+// 2. Prepare Snapshots
+// 3. Set QuotaID for upperDir and workDir
+// 4. Stop all containers and alidocker.
+// 5. mv oldUpperDir/* => upperDir/
+// 6. Convert oldContainerMeta to PouchContainer container metaJSON
+// 7. Stop containerd
+// 8. Install pouch
+// 9. Start all container
+
+// coldMigrator is a tool to migrate docker containers to pouch containers,
+// which we should first stop all containers before migration.
+type coldMigrator struct {
+	// dockerd home dir
+	dockerHomeDir string
+
+	// store map of old UpperDir and new UpperDir
+	upperDirMappingList []*upperDirMapping
+	// store all containers info
+	allContainers map[string]bool
+	// store all running containers
+	runningContainers []string
+	// store all images using by containers
+	images map[string]struct{}
+}
+
+// upperDirMapping stores overlayfs upperDir map for docker and pouch.
+type upperDirMapping struct {
+	// specify docker UpperDir
+	srcDir string
+	// specify pouch UpperDir
+	dstDir string
+}
+
+// NewColdMigrator creates a migrator tool instance.
+func NewColdMigrator(cfg Config) (Migrator, error) {
+	migrator := &coldMigrator{
+		dockerHomeDir: cfg.DockerHomeDir,
+		allContainers: map[string]bool{},
+	}
+
+	return migrator, nil
+}
+
+// PreMigrate prepares things for migration
+// * pull image to pouch
+// * create snapshot for container
+// * set snapshot upperDir, workDir diskquota
+// * convert docker container metaJSON to pouch container metaJSON
+func (cm *coldMigrator) PreMigrate(ctx context.Context, dockerCli *docker.Dockerd) error {
+	// Get all docker containers on host.
+	containers, err := dockerCli.ContainerList()
+	if err != nil {
+		return fmt.Errorf("failed to get containers list: %v", err)
+	}
+	logrus.Debugf("Get %d containers", len(containers))
+
+	if len(containers) == 0 {
+		logrus.Info("Empty host, no need covert containers")
+		return nil
+	}
+
+	// Get all volumes on host.
+	dockerVolumes, err := dockerCli.VolumeList()
+	if err != nil {
+		return fmt.Errorf("failed to get volumes list: %v", err)
+	}
+	pouchVolumes, err := convertor.ToVolumes(dockerVolumes)
+	if err != nil {
+		return err
+	}
+
+	var (
+		// volumeRefs to count volume references
+		volumeRefs    = map[string]string{}
+		pouchHomeDir  = getPouchHomeDir(cm.dockerHomeDir)
```

Comment:
`getPouchHomeDir` could be treated as a convertor too. what about moving it to `convertor.ToPouchHomeDir`?
what's more, why don't you just store the pouchHomeDir inside the migrator, it is unnecessary to generate every time....

Reply:
`pouchHomeDir` can get from `dockerHomeDir`, so I think no need to store this parameters

Reply:
`getPouchHomeDir could be treated as a convertor too. what about moving it to convertor.ToPouchHomeDir?`

`convertor` is used to convert the container meta json, not the config file. So I don't think we should mv the `getPouchHomeDir` to `convertor.ToPouchHomeDir`. WDYT?